py_bentoctl/auth_helper.py

```diff
@@ -96,12 +106,119 @@ def fetch_existing_client_id(token: str, client_id: str) -> Optional[str]:
+    if existing_role := fetch_existing_client_role(token, client_id, role_name):
+        return existing_role
+
+    # Create client role if needed
+    client_roles_endpoint = f"{KC_ADMIN_API_CLIENTS_ENDPOINT}/{client_id}/roles"
+    res = keycloak_req(client_roles_endpoint, bearer_token=token, method="post", json={
+        "clientRole": True,
+        "name": role_name,
+    })
+
+    if not res.ok:
+        err(f"    Failed to create {client_id} client role : {role_name}; {res.status_code}")
+        exit(1)
+
+    # role creation response returns no data, fetch the created RoleRepresentation for later use
+    created_role_rep = fetch_existing_client_role(token, client_id, role_name, verbose=False)
+    warn(f"    Created client role: {role_name}.")
+    return created_role_rep
+
+
+def create_group_or_exit(token: str, group_rep: dict, parent_group_rep: dict = None) -> Optional[dict]:
+    # try to get existing group first
+    if existing_group := fetch_existing_group_rep_or_exit(token, group_rep["name"], parent_group_rep):
+        return existing_group
+
+    # group creation endpoint
+    group_endpoint = KC_ADMIN_API_GROUP_ENDPOINT
+    if parent_group_rep:
+        # use sub-group creation endpoint if a parent group is passed
+        group_endpoint = f"{group_endpoint}/{parent_group_rep['id']}/children"
+
+    res = keycloak_req(f"{group_endpoint}", bearer_token=token, method="post", json=group_rep)
+    if not res.ok:
+        err(f"    Failed to create group: {group_rep}; {res.status_code}")
+        exit(1)
+
+    # group creation response returns no data, fetch the created GroupRepresentation for later use
+    created_group = fetch_existing_group_rep_or_exit(token, group_rep["name"], parent_group_rep, verbose=False)
+
+    warn(f"    Created group: {created_group['path']}")
+    return created_group
+
+
+def add_client_role_mapping_to_group_or_exit(token: str, group_rep: dict, client_id: str, role_rep: dict) -> None:
+    role_mappings_endpoint = f"{KC_ADMIN_API_GROUP_ENDPOINT}/{group_rep['id']}/role-mappings/clients/{client_id}"
+    existing_mappings_res = keycloak_req(role_mappings_endpoint, bearer_token=token)
+    if existing_mappings_res.ok:
+        target_role_name = role_rep["name"]
+        for role_map in existing_mappings_res.json():
+            if target_role_name == role_map["name"]:
+                warn(f"    Found existing client-level group role: {group_rep['path']}; using that.")
+                return
+
+    # create client role-mapping for given group
+    client_res = keycloak_req(
+        role_mappings_endpoint,
+        method="post",
+        bearer_token=token,
+        data=json.dumps([role_rep])   # RoleRepresentation needs to be in an array and sent as data
+    )
+    if not client_res.ok:
+        err(f"    Failed to add client-level role {role_rep['name']} to group {group_rep['path']}")
+        exit(1)
+
+    warn(f"    Created client-level role mapping for group: {group_rep['path']}")
```

Comment:
is this a warning? other places use a created + done approach

Reply:
true, changed the one for client roles scopes too